### PR TITLE
feat(l10n): Split payments-server single FTL file into per-component …

### PR DESCRIPTION
**Pull request by vpomerleau**
## Because

- Organizing l10n strings in per-component files aligns payments-server with the setup used for auth-server and settings. Splitting the FTL files also improves maintability by more tightling coupling the strings with the components where the strings are used.
- Changing the default locale from en-US to en aligns with other fxa packages.
- Changing the concatenated file name from main.ftl to payments.ftl improve file name clarity.

## This pull request

* Create a branding file in **/src (updated Nov 2)**
* Create a gruntfile with tasks for FTL concatenation
* Rename the destination file from main.ftl to payments.ftl
* Switch payments-server's default locale from en-US to en
* Create individual FTL files per component and move messages to their respective component FTL file
* Add a l10n entry in the package readme

Update: Nov 2, 2022

* Update clone-l10n.sh to copy 'payments' ftl files instead of 'main'
* Update AppLocalizationProvider to use 'payments' bundles

Update: Nov 4, 2022

* Replace setupFluentLocalizationTest with getFtlBundle/getFtlFromPackage from fxa-react
* Remove setupFluentLocalizationTest function, no longer in use
* Add merge-ftl:test task to package.json test:frontend script
* Remove legacy strings from FTL files
* Update currency and date formats to use 'en' as default locale

## Issue that this pull request solves

Closes: #FXA-5996

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).

### CODE REVIEW

File: packages/fxa-payments-server/Gruntfile.js
Comment (vpomerleau) [inline packages/fxa-payments-server/Gruntfile.js:22]: l10n testing has not yet been added for payments-server

File: packages/fxa-payments-server/public/locales/en-US/unused-strings.ftl
Comment (vpomerleau) [inline packages/fxa-payments-server/public/locales/en-US/unused-strings.ftl:1]: Could not locate where these strings are used. I'm not sure if these strings are deprecated.

This ftl file is not currently concatenated into the l10n package.

File: packages/fxa-payments-server/src/lib/errors.ts
Comment (vpomerleau) [inline packages/fxa-payments-server/src/lib/errors.ts:44]: Updated to match ID used in FTL file

**Comment by LZoog** [comment]
@bcolsson This PR is the one that renames `main.ftl` to `payments.ftl`. I can ping you in Slack when this is ready so we can all coordinate on when to merge this, you'd previously said...

"This would require some coordination with Pontoon engineering. Renaming/moving a file in the repo keeps the existing translations, but we would need to run some commands on the Pontoon database to preserve things like translation history / translation attribution / pending suggestions."

@vpomerleau We will also want to make a couple of tweaks on the l10n side. We'll want to add a payments grunt command [here](https://github.com/mozilla/fxa-content-server-l10n/blob/main/.github/workflows/l10n_extract.yaml#L40) as well as updating `en-US` to `en` [here](https://github.com/mozilla/fxa-content-server-l10n/blob/main/scripts/extract_strings.sh#L120). You can open a draft PR at any time over there and reference this PR, we'll want to merge that right after merging this over here.

File: packages/fxa-payments-server/Gruntfile.js
Comment (LZoog) [inline packages/fxa-payments-server/Gruntfile.js:17]: Nitpick: there may be a way to reference `'src/lib/l10n/branding.ftl'` first and then reference `src/**/*.ftl` while excluding the branding file to avoid this many paths. However this doesn't really matter since we'll eventually change this to `../fxa-shared/lib/l10n/branding.ftl', 'src/**/*.ftl'`.

Might also be good to leave a comment above `branding.ftl` (and not in the file since that'd get into the l10n repo) noting that the branding file is temporary until we can use the one from `fxa-shared`. Haven't filed an issue for that yet, I'll comment back here when I have (may be before or after this is merged).

File: packages/fxa-payments-server/Gruntfile.js
Comment (LZoog) [inline packages/fxa-payments-server/Gruntfile.js:47]: For this to watch properly, you have to update the `pm2` config file, which contains these kind of watch processes.

**Comment by vpomerleau** [comment]
> @vpomerleau We will also want to make a couple of tweaks on the l10n side. We'll want to add a payments grunt command [here](https://github.com/mozilla/fxa-content-server-l10n/blob/main/.github/workflows/l10n_extract.yaml#L40) as well as updating `en-US` to `en` [here](https://github.com/mozilla/fxa-content-server-l10n/blob/main/scripts/extract_strings.sh#L120). You can open a draft PR at any time over there and reference this PR, we'll want to merge that right after merging this over here.

Draft PR on l10n side is now open: [PR-657](https://github.com/mozilla/fxa-content-server-l10n/pull/657)

File: packages/fxa-payments-server/Gruntfile.js
Comment (vpomerleau) [reply packages/fxa-payments-server/Gruntfile.js:17]: I agree, it's not as clean as I would like. I could do

```
concat: {
  ftl: {
    src: [
      // 'src/branding.ftl' is temporary and will be replaced
      //  with '../fxa-shared/lib/l10n/branding.ftl' in a later ticket
      //  This will require coordination with l10n to resolve
      // conflicting IDs for identical terms.
      'src/branding.ftl',
      'src/**/*.ftl',
    ],
    dest: 'public/locales/en/payments.ftl',
  },
```

I've tested with `yarn run merge-ftl` and this only includes one instance of branding.ftl (concatenated first) in the destination file.

Downside is that the strings shared by multiple components are included midway in the concatenated file rather than the top. Might not matter?

File: packages/fxa-payments-server/Gruntfile.js
Comment (vpomerleau) [reply packages/fxa-payments-server/Gruntfile.js:47]: Thanks for pointing that out, I'll include that in the updated commit.

**Comment by bcolsson** [comment]
> @bcolsson This PR is the one that renames `main.ftl` to `payments.ftl`. I can ping you in Slack when this is ready so we can all coordinate on when to merge this, you'd previously said...
> 
> > "This would require some coordination with Pontoon engineering. Renaming/moving a file in the repo keeps the existing translations, but we would need to run some commands on the Pontoon database to preserve things like translation history / translation attribution / pending suggestions."
> 

@LZoog - thanks for the heads up. I know this is still WIP and I may not be reading the code correctly so please correct me if I'm wrong, but this appears to be outputting both main.ftl and payments.ftl. Am I correct in assuming in the final shape of this PR will result in main.ftl no longer being created? Otherwise we'll end up having two files with duplicate strings being sent to localizers.

File: packages/fxa-payments-server/Gruntfile.js
Comment (LZoog) [reply packages/fxa-payments-server/Gruntfile.js:17]: Nice, that looks cleaner. I think we're going to end up going with `src/**/*.ftl` anyway once we swap out the branding file, so that seems great.

> Downside is that the strings shared by multiple components are included midway in the concatenated file rather than the top. Might not matter?

I think this is fine. It might "feel" more ideal to be first under "branding" but doesn't affect anything so long as the comments match up and make sense once concatenated.

**Comment by LZoog** [comment]
> > @bcolsson This PR is the one that renames `main.ftl` to `payments.ftl`. I can ping you in Slack when this is ready so we can all coordinate on when to merge this, you'd previously said...
> > > "This would require some coordination with Pontoon engineering. Renaming/moving a file in the repo keeps the existing translations, but we would need to run some commands on the Pontoon database to preserve things like translation history / translation attribution / pending suggestions."
> 
> @LZoog - thanks for the heads up. I know this is still WIP and I may not be reading the code correctly so please correct me if I'm wrong, but this appears to be outputting both main.ftl and payments.ftl. Am I correct in assuming in the final shape of this PR will result in main.ftl no longer being created? Otherwise we'll end up having two files with duplicate strings being sent to localizers.

That is correct, there will only be `payments.ftl` after this is finished.

@vpomerleau you'll want to modify the `.gitignore` file to not commit anything in `packages/fxa-payments-server/public/locales`, delete `main.ftl`, and make sure the new `payments.ftl` file isn't committed. You might need to `git rm --cached`.

**Comment by bcolsson** [comment]
@LZoog - Also just FYI, we probably won't be able to handle the Pontoon side of things this week, but if the PR can be finalized ahead of next week we should be able to do the migration then.

File: packages/fxa-payments-server/src/en.ftl
Comment (vpomerleau) [inline packages/fxa-payments-server/src/en.ftl:13]: Strings from line 13 onwards don't seem to be used in any routes/components. Are these safe to delete?

Comment (sardesam) [reply packages/fxa-payments-server/src/en.ftl:13]: 👍 Confirmed that they are all legacy strings and can be removed.

File: packages/fxa-payments-server/Gruntfile.js
Comment (LZoog) [reply packages/fxa-payments-server/Gruntfile.js:23]: Payments-server is not yet using the l10n test utilities we've begun to define in `fxa-react`, but it does have some l10n testing with [`setupFluentLocalizationTest`](https://github.com/mozilla/fxa/blob/81d18892e22c1e0e1efec96b1270251bbf3e92f2/packages/fxa-payments-server/src/lib/test-utils.tsx#L695) and `getLocalizedMessage`.

If we leave `setupFluentLocalizationTest` as-is, it's going to attempt to pull from the `public/locales` dir, and it'll test against the l10n side's latest and not our latest. We need to set up that `merge-ftl:test` script in `package.json` and have it run with the test command, and then we should remove `setupFluentLocalizationTest` in favor of using `fxa-react`'s `getFtlFromPackage`. There's [a comment](https://github.com/mozilla/fxa/blob/main/packages/fxa-react/lib/test-utils/index.ts#L45) in that file about doing this in FXA-5996. You'll need to update any function needing this to be `async`.

I think we can leave `getLocalizedMessage` for now, and have other issues filed for the new testing methods after this and FXA-6001 are merged.

File: packages/fxa-payments-server/README.md
Comment (LZoog) [inline packages/fxa-payments-server/README.md:89]: I know we have this in `auth-server` too, but `extract-and-pull-request.sh` is actually no longer a script, it's a GH action. We do have `extract_strings.sh` over there though... I'll go through our docs later when we do doc tickets. 🤷‍♀️

**Review by LZoog — COMMENTED** [review]
Looking good so far. I need to test it locally but left you a couple initial comments.

File: packages/fxa-payments-server/src/components/en.ftl
Comment (vpomerleau) [reply packages/fxa-payments-server/src/components/en.ftl:1]: Yeah, you're probably right that this level of detail is not needed. I'll move all shared strings to `src/en.ftl` with a group-level comment "Strings shared by multiple components/routes"

File: packages/fxa-payments-server/README.md
Comment (vpomerleau) [reply packages/fxa-payments-server/README.md:89]: Should I entirely remove this bit?

> and the `extract-and-pull-request.sh` script that runs in `fxa-content-server-l10n` on a weekly cadence.

File: packages/fxa-payments-server/Gruntfile.js
Comment (LZoog) [reply packages/fxa-payments-server/Gruntfile.js:23]: @vpomerleau Your changes look great, we still need to add the file to `.gitignore` though; we don't want to commit `test/payments.ftl`. We should also be able to remove `setupFluentLocalizationTest`.

File: packages/fxa-payments-server/src/lib/formats.ts
Comment (LZoog) [inline packages/fxa-payments-server/src/lib/formats.ts:47]: @StaberindeZA This is definitely one I'd want SubPlat to check on. 🤔

File: packages/fxa-payments-server/README.md
Comment (LZoog) [reply packages/fxa-payments-server/README.md:89]: Eh, either way actually since I guess a GH action is technically a script and this'll be tweaked in other tickets anyway, I just think of `.sh` files when I see `script`.

Oh, we can update that link though. This should go to the deployed ecosystem-platform docs, not `localhost:3000` (this was probably my mistake in auth-server's readme, lol).

File: packages/fxa-payments-server/Gruntfile.js
Comment (vpomerleau) [reply packages/fxa-payments-server/Gruntfile.js:23]: Right, I'll clean that `test` folder from the commit. 
I'd removed `setupFluentLocalizationTest` already in a small extra push, probably just as you were reviewing :)

File: packages/fxa-payments-server/README.md
Comment (vpomerleau) [reply packages/fxa-payments-server/README.md:89]: I removed the script name and updated the link

File: packages/fxa-payments-server/Gruntfile.js
Comment (LZoog) [reply packages/fxa-payments-server/Gruntfile.js:23]: @vpomerleau I think we still need to add the test directory to the `.gitignore`.

Comment (vpomerleau) [reply packages/fxa-payments-server/Gruntfile.js:23]: There was also a .gitignore template in the /src folder - I've removed that one.

File: packages/fxa-payments-server/package.json
Comment (LZoog) [inline packages/fxa-payments-server/package.json:15]: I believe we want to run `grunt merge-ftl` on `start` similar to settings; having just the watch command running doesn't run it the command right away on `yarn start` so if strings have changed between `yarn stop` and `yarn start` it doesn't update. (I don't think we need it in `postinstall` like it is in settings though, since it gets overwritten, that's probably something we can remove on the settings side)

**Review by LZoog — COMMENTED** [review]
LGTM with a couple comments, I'll circle back tomorrow and r+ after those have been addressed and then once SubPlat and @bcolsson r+, I'll pull the commit into another PR and once CI goes green, I'm comfortable merging. 🥳

File: packages/fxa-payments-server/Gruntfile.js
Comment (vpomerleau) [reply packages/fxa-payments-server/Gruntfile.js:23]: @LZoog Most are covered in the global `gitignore`. I'll admit I remove this a bit too fast - here are the lines that weren't in the global:
```
.env.local
.env.development.local
.env.test.local
.env.production.local
/.pnp
.pnp.js
```
What would you think if I move these to global and still remove the file in `/src`?

Comment (vpomerleau) [reply packages/fxa-payments-server/Gruntfile.js:23]: I've addressed the edits from [FXA-6255](https://mozilla-hub.atlassian.net/browse/FXA-6255) in my latest commit.

File: packages/fxa-payments-server/public/locales/en-US/main.ftl
Comment (bcolsson) [reply packages/fxa-payments-server/public/locales/en-US/main.ftl:1]: Thanks @vpomerleau.  I checked the output of your commit in combination with the associated [workflow update ](https://github.com/mozilla/fxa-content-server-l10n/pull/657) on `fxa-content-server-l10n` with moz-l10n-lint / compare-locales and not seeing any issues. Did a quick spot check manually as well but nothing stands out. I'll approve the review.

@LZoog  - Before merging these changes, we'll need to coordinate a time once the final PRs are ready. We'll use those PRs to do a dry run migration for Pontoon.

**Review by bcolsson — APPROVED** [review]

**Comment by vpomerleau** [comment]
Latest push: Added the following new FTL ID-message pairs after diff check of `main.ftl` between the initial commit (2b76e4712086def4c93fbb896950b2126a239551) from this branch with what's currently on `main`.

```
brand-name-firefox-logo = { -brand-name-firefox } logo
subscription-iaperrorupgrade-title = We can’t upgrade you quite yet
```

File: packages/fxa-payments-server/public/locales/en-US/unused-strings.ftl
Comment (StaberindeZA) [reply packages/fxa-payments-server/public/locales/en-US/unused-strings.ftl]: @vpomerleau are there still unused strings? I don't see any strings commented out at the bottom of `src/ftl`

Comment (StaberindeZA) [reply packages/fxa-payments-server/public/locales/en-US/unused-strings.ftl]: Ah never mind. I saw your conversation with @sardesam. Apologies for the double ping.

File: packages/fxa-payments-server/public/locales/en-US/main.ftl
Comment (StaberindeZA) [reply packages/fxa-payments-server/public/locales/en-US/main.ftl:1]: @vpomerleau @LZoog Do we happen to have the list of unused strings available somewhere still?

Comment (vpomerleau) [reply packages/fxa-payments-server/public/locales/en-US/main.ftl:1]: @StaberindeZA Here's a copy, please expand:

<details>
<summary>Unused Strings</summary>

```
# UNUSED STRINGS?

# # $productName (String) - The name of the subscribed product.
# # $amount (Number) - The amount billed. It will be formatted as currency.

# # $intervalCount (Number) - The interval between payments, in days.
# day-based-plan-details-amount = { $intervalCount ->
#   [one] { $productName } billed { $amount } daily
#   *[other] { $productName } billed { $amount } every { $intervalCount } days
# }
# # $intervalCount (Number) - The interval between payments, in weeks.
# week-based-plan-details-amount = { $intervalCount ->
#   [one] { $productName } billed { $amount } weekly
#   *[other] { $productName } billed { $amount } every { $intervalCount } weeks
# }
# # $intervalCount (Number) - The interval between payments, in months.
# month-based-plan-details-amount = { $intervalCount ->
#   [one] { $productName } billed { $amount } monthly
#   *[other] { $productName } billed { $amount } every { $intervalCount } months
# }
# # $intervalCount (Number) - The interval between payments, in years.
# year-based-plan-details-amount = { $intervalCount ->
#   [one] { $productName } billed { $amount } yearly
#   *[other] { $productName } billed { $amount } every { $intervalCount } years
# }
# ## Payment form

# payment-ccn =
#   .label = Card number
# payment-exp =
#   .label = Expiration
# payment-cvc =
#   .label = CVC
# payment-zip =
#   .label = ZIP code
# payment-validate-zip-required = Zip code is required
# payment-validate-zip-short = Zip code is too short
# ## Subscription redirect

# sub-redirect-ready = Your subscription is ready
# sub-redirect-copy = Please take a moment to tell us about your experience.
# sub-redirect-skip-survey = No thanks, just take me to my product.
# ## Subscription upgrade

# sub-update-card-exp = Expires { $cardExpMonth }/{ $cardExpYear }
# sub-change-indicator =
#   .aria-label = change indicator
# ##

# account-activated = Your account is activated, <userEl/>
# ## Payment confirmation

# payment-confirmation-alert = Click here to download
# payment-confirmation-mobile-alert = Didn’t open app? <a>Click Here</a>
# payment-confirmation-billing-heading = Billed to
# ## Coupon
# coupon-discount = Discount
# coupon-discount-applied = Discount Reward Applied
# # Subscriptions - Reactivate
# reactivate-panel-date = You cancelled your subscription on { $date }.
# ## Subscription item

# sub-item-missing = Problem loading subscriptions
# sub-item-missing-msg = Please try again later.
# ## Subscriptions error
# sub-subscription-error =
#   .title = Problem loading subscriptions
# # Routes
# pay-with-heading-other = Select payment option
```
</details>

File: packages/fxa-payments-server/src/lib/formats.ts
Comment (StaberindeZA) [reply packages/fxa-payments-server/src/lib/formats.ts:47]: Looks good to me. We only use it for fallback text, and it seems to still work as expected. Thank you!

**Review by StaberindeZA — APPROVED** [review]
LGTM! Thank you for all these changes. Super happy to have this change in.

I believe I responded to all pings to me, but if not, please let me know.

**Comment by LZoog** [comment]
Okay, pulling the merge button here. Still not sure why CI is seeing this as a fork and not allowing circle token access per our docs, but that's all that's failing so, 🎉

Thank you @vpomerleau for all your work on this PR! 😍

**Pull request merged**
